cmd/mysql-state-monitor/main.go

```diff
@@ -0,0 +1,123 @@
+package main
+
+import (
+	"log"
+	"net"
+	"os"
+	"os/signal"
+	"strings"
+)
+
+type MySQLState string
+
+const (
+	MySQLReady   MySQLState = "ready"
+	MySQLDown    MySQLState = "down"
+	MySQLStartup MySQLState = "startup"
+	MySQLUnknown MySQLState = "unknown"
+)
+
+func parseDatum(datum string) MySQLState {
+	lines := strings.Split(datum, "\n")
+
+	if lines[0] == "READY=1" {
+		return MySQLReady
+	}
+
+	if lines[0] == "STOPPING=1" {
+		return MySQLDown
+	}
+
+	if strings.HasPrefix(lines[0], "STATUS=") {
+		status := strings.TrimPrefix(lines[0], "STATUS=")
+		switch status {
+		case "Server is operational":
+			return MySQLReady
+		case "Server shutdown in progress":
+			return MySQLDown
+		case "Server startup in progress",
+			"Data Dictionary upgrade in progress",
+			"Data Dictionary upgrade complete",
+			"Server upgrade in progress",
+			"Server upgrade complete",
+			"Server downgrade in progress",
+			"Server downgrade complete",
+			"Data Dictionary upgrade from MySQL 5.7 in progress",
+			"Data Dictionary upgrade from MySQL 5.7 complete",
+			"Server shutdown complete": // we treat this as startup because during init, MySQL notifies this even if it's up
+			return MySQLStartup
+		}
+	}
+
+	return MySQLUnknown
+}
+
+func main() {
+	log.Println("Starting mysql-state-monitor")
+
+	socketPath, envDefined := os.LookupEnv("NOTIFY_SOCKET")
+	if !envDefined {
+		log.Fatalln("NOTIFY_SOCKET env variable is required")
+	}
+
+	stateFilePath, envDefined := os.LookupEnv("MYSQL_STATE_FILE")
+	if !envDefined {
+		log.Fatalln("MYSQL_STATE_FILE env variable is required")
+	}
+
+	stateFile, err := os.Create(stateFilePath)
+	if err != nil {
+		log.Fatalf("Failed to create state file: %s", err)
+	}
+
+	if _, err := os.Stat(socketPath); err == nil {
+		if err := os.Remove(socketPath); err != nil {
+			log.Fatalf("Failed to remove %s: %s", socketPath, err)
+		}
+	}
+
+	addr, err := net.ResolveUnixAddr("unixgram", socketPath)
+	if err != nil {
+		log.Fatalf("Failed to resolve unix addr %s: %s", socketPath, err)
+	}
+
+	conn, err := net.ListenUnixgram("unixgram", addr)
+	if err != nil {
+		log.Fatalf("Failed to listen unixgram %s: %s", socketPath, err)
+	}
+	defer conn.Close()
```

Comment:
if there's an error returned there, conn is not opened. I don't think we should defer before err check

Reply:
But regardless when defer is not called after `os.Exit` the connection will still be released since the application is terminated and the OS will release all the resources.

But like @egegunes it is also most probable that the connection is not created if we encounter an error here.

Reply:
I meant that we use `log.Fatal` after using defer. I suggest either deleting the `defer` statement or moving all the logic from `main` to another function that returns an error, which will make sense to use `defer`